clang/lib/Sema/SemaBounds.cpp

```diff
@@ -2712,34 +2705,43 @@ namespace {
       }
    }
 
-   void UpdateWidenedBounds(BoundsAnalysis &BA, const CFGBlock *Block,
-                            CheckingState &State) {
-     for (const auto item : BA.GetWidenedBounds(Block)) {
-       const VarDecl *V = item.first;
-       BoundsExpr *Bounds = item.second;
-
-       // BoundsAnalysis currently uses VarDecls as keys in the widened
-       // bounds data structure, so we create an AbstractSet for each 
-       // VarDecl in the widened bounds. TODO: use AbstractSets as keys
-       // in BoundsAnalysis (checkedc-clang issue #1015).
-       const AbstractSet *A = AbstractSetMgr.GetOrCreateAbstractSet(V);
-       auto I = State.ObservedBounds.find(A);
-       if (I != State.ObservedBounds.end())
-         I->second = Bounds;
+   void UpdateWidenedBounds(BoundsWideningAnalysis &BWA, const CFGBlock *Block,
+                            Stmt *CurrStmt, CheckingState &State) {
+
+     // Get the bounds widened before the current statement.
+     BoundsMapTy WidenedBounds = BWA.GetStmtIn(Block, CurrStmt);
+
+     // If a variable does not exist in WidenedBounds but exists in
+     // State.ObservedBounds it means that the bounds of the variable got
+     // killed before the current statement. So we reset the bounds of the
+     // variable in State.ObservedBounds to its declared bounds.
+     for (auto I = State.ObservedBounds.begin(),
+               E = State.ObservedBounds.end(); I != E; ++I) {
+       const AbstractSet *A = I->first;
+       const VarDecl *V = A->GetVarDecl();
+       if (!V)
+         continue;
+
+       if (WidenedBounds.find(V) == WidenedBounds.end())
+         I->second = S.NormalizeBounds(V);
      }
-   }
 
-   void ResetKilledBounds(BoundsAnalysis &BA, const CFGBlock *Block,
-                          const Stmt *St, CheckingState &State) {
-     for (const VarDecl *V : BA.GetKilledBounds(Block, St)) {
-       // BoundsAnalysis currently uses VarDecls as keys in the killed
-       // bounds data structure, so we create an AbstractSet for each
-       // VarDecl in the killed bounds. TODO: use AbstractSets as keys
-       // in BoundsAnalysis (checkedc-clang issue #1015).
+     // Update State.ObservedBounds with the WidenedBounds before the current
+     // statement.
+     for (const auto VarBoundsPair : WidenedBounds) {
+       const VarDecl *V = VarBoundsPair.first;
+       BoundsExpr *Bounds = VarBoundsPair.second;
+       if (!Bounds)
+         continue;
+
+       // BoundsWideningAnalysis currently uses VarDecls as keys in the widened
+       // bounds data structure, so we create an AbstractSet for each VarDecl
```

Comment:
Minor correction but will help clarify things for a reader unfamiliar with all the details: ..... so we **get** the `AbstractSet` for each `VarDecl` .....